Retrofit the release workflow to be useful for doing a pre-release

ariya opened the pull request:
Once this PR is merged, there will be a manually-triggered workflow which does the following, for each OSS and EE edition, to facilitate testing both the Uberjars and container images (before a release):

1. Build Uberjars, stash the Uberjars as downloadable artifacts
2. Check the health check when launching each Uberjar with Java 11 and 17
3. Put the Uberjar in the container image, push it to DockerHub
4. Verify the image by doing `docker pull`

Step 3 and 4 are carried out in isolation, thereby allowing the full round-trip (`docker push` and `docker pull`) verification.

Testing this is easier done in a fork (e.g. in my case, [github.com/ariya/metabase](https://github.com/ariya/metabase)). The workflow can be triggered by supplying the commit hash which we want to check:

![image](https://user-images.githubusercontent.com/7288/202596636-efe9dcef-aff6-4f91-b92f-c57de85a65d2.png)

The visualization of all the jobs being executed thereafter:

![image](https://user-images.githubusercontent.com/7288/202596755-1510033b-482c-4da2-b49e-744cecaeaa4f.png)

The container image is also available in Docker Hub (shown here only for `metabase-enterprise`, but `metabase` the OSS edition is also available):

![image](https://user-images.githubusercontent.com/7288/202597011-96ded177-bc73-4fc1-968a-a7a73b35c16b.png)

Commit:
Retrofit the release workflow to be useful for doing a pre-release

ariya commented on .github/workflows/pre-release.yml:
```diff
@@ -27,8 +21,8 @@ jobs:
-      with:
-        ref: ${{ github.event.inputs.commit }}
+    - name: Switch to commit ${{ github.event.inputs.commit }}
+      run: git checkout ${{ github.event.inputs.commit }}
```
Manual commit switch seems to be more reliable, based on my testing.

nemanjaglumac replied:
More reliable in what sense?
Did you notice failures when using actions/checkout?

ariya replied:
I tried using `ref` again, it consistently failed on me (not sure what I'm missing here):

![image](https://user-images.githubusercontent.com/7288/202768208-7dc4bebf-a4fa-4f7a-a017-03af6ebf42d2.png)

nemanjaglumac replied:
Oh, it **might be** because you're not supplying the full SHA?
In all of my testing, I was always copying the full SHA and used it to trigger the workflow.

This is of secondary importance.
I'm happy to use either.

nemanjaglumac replied:
IF that's the reason why, maybe it's not even bad to use that mechanism as an enforcing function because the SHA itself is going to be used in an image name.

That way, we can avoid having multiple images for the same commit. E.g.
- `metabase/metabase: 3e518549f803ff050fe6ae43ddb6a003908b18a0`
- `metabase/metabase: 3e5185`
- `metabase/metabase: 3e518549f803ff050fe6ae43ddb6a0039`

nemanjaglumac replied:
^ Either that or we need to have an additional logic that validates the SHA and possibly truncates it to a minimum required length, regardless of the input provided?

ariya replied:
That did the trick, thank you!
It's very strange since the canonical `git` can work with any length of SHA1 (as long as it's unique).

ariya replied:
I like the idea of limiting the length of the SHA1 hash to just 8 characters (more than sufficient, for a while).

nemanjaglumac replied:
^ https://github.com/metabase/metabase/pull/26597/commits/0e73ec3699f7e32c8e32338d6d6c78106accc3a4 should do the trick.

And I switched back to implicitly enforcing full SHA-1 as an input in https://github.com/metabase/metabase/pull/26597/commits/3039a30d544564ce0720f8a709549de9858c67d4.

Feel free to change/adapt any part of it as you see fit.

ariya commented on .github/workflows/pre-release.yml:
```diff
@@ -95,47 +83,75 @@ jobs:
     - name: Set up Docker Buildx
       id: buildx
-      uses: docker/setup-buildx-action@v1
+      uses: docker/setup-buildx-action@v2
```
Bump the action version, to avoid the nagging of outdated Node.js v12 action.

ariya commented on .github/workflows/pre-release.yml:
```diff
@@ -95,47 +83,75 @@ jobs:
         driver-opts: network=host
     - name: Build ${{ matrix.edition }} container
-      uses: docker/build-push-action@v2
+      uses: docker/build-push-action@v3
```
idem

ariya commented on .github/workflows/pre-release.yml:
```diff
@@ -95,47 +83,75 @@ jobs:
-          fi
+        if [[ ${{ matrix.edition }} == ee ]]; then
+          echo "Metabase EE: image is going to be pushed to ${{ github.repository_owner }}/metabase-enterprise:${{ github.event.inputs.commit }}"
+          echo "DOCKERHUB_REPO=${{ github.repository_owner }}/metabase-enterprise:${{ github.event.inputs.commit }}" >> $GITHUB_ENV
```
`github.repository_owner` is used here so that things are also working correctly in a fork (and therefore permits easy testing of this commit/PR).

nemanjaglumac reviewed (approved):
Works for me :)
Great and thanks for working on this!

nemanjaglumac reviewed (changes requested):
Actually, wait.
There are a few improvement that we can still make.

Sorry for going back and forth but I didn't notice something originally.

nemanjaglumac commented on .github/workflows/pre-release.yml:
```diff
@@ -95,47 +83,75 @@ jobs:
+        docker push ${{ env.DOCKERHUB_REPO }}
+        echo "Finished!"
+
+  verify-docker-pull:
```
WDYT if we have `verify-docker-pull` only as a **step** in a previous job, rather than having to spin up new container for a completely new **job**?

That way we can save some resources and we can re-use the conditional logic re: `DOCKERHUB_REPO`. That could dry the code up significantly.

ariya replied:
If we do that, the `docker pull` will only pull from its local cache (since the container image was constructed a few seconds) and not from Dockerhub.
The reason for a separate job is so that the `push` and `pull` are isolated (they don't share the execution environment).

nemanjaglumac replied:
And environment variables are not persisted between jobs, right?

ariya replied:
The env variables can be scoped to the entire workflow only if they are defined statically at the top level.

TBH I personally wouldn't worry much about duplication for now, since at some point it can be abstracted away (the case with our custom actions e.g. `prepare-frontend` etc).

nemanjaglumac committed:
Distinguish between `metabase` main repo and the custom ones

nemanjaglumac reviewed (approved):
Approving (again) just to unblock the PR.
But, since I ended up contributing, please wait for at least one more reviewer to approve before merging.

nemanjaglumac committed:
Truncate commit id to specific hash length

We need to make sure all images are tagged with exactly the same hash length
in order to avoid having multiple images for the same commit.

nemanjaglumac committed:
Enforce full commit SHA-1 hash is used for an input

`actions/checkout@v3` will fail if full SHA-1 hash is not provided.
We want to use this implicit mechanism as an enforcing function
for uniform manual inputs.

nemanjaglumac committed:
Never run script from the main repo

- This greatly simplifies things.
- The script is meant to be run from forks of Metabase for testing purposes and/or for custom builds.

nemanjaglumac committed:
Safeguard against missing `CUSTOM_RELEASE_REPO`
